microsoft/PowerToys
# [FancyZones] Reset zone settings after restart fix.

## Summary of the Pull Request

**What is this about:**

For some users zone settings are reset to default on every restart. 
The virtual desktop id that is part of the `device-id` is changed on every restart until the user doesn't create another virtual desktop (even if it was deleted immediately). After the creation of the second VD, their ids are saved in the registry and would be the same on every restart.

**What is include in the PR:** 

* Replaced `device-id` string with a structure.
* Changed virtual desktop update handle.
* Updated obtain of the virtual desktop ids.
* Synchronize virtual desktop ids in `app-zone-history` and `zones-settings` maps with current virtual desktop ids. In case if ids are saved in the registry, save them in the settings as it is, otherwise save with the default `{00000000-0000-0000-0000-000000000000}` id. While using, the default id would be replaced with a current VD id.

**How does someone test / validate:** 

* Verify that every restart layouts are the same and not reset with the default priority grid.
* Verify that the newly created virtual desktop has the same layouts as on the previous virtual desktop.
* After changing the layout on one of several virtual desktops verify that all other stays the same as they were before changes.

Check this behavior with and without saved virtual desktop ids in the registry.
To remove ids from the registry, remove the 
* `CurrentVirtualDesktop` from `\HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Explorer\SessionInfo\1\VirtualDesktops`
* `VirtualDesktopIDs` from `\HKEY_CURRENT_USER\Software\Microsoft\Windows\CurrentVersion\Explorer\VirtualDesktops`

## Quality Checklist

- [x] **Linked issue:** #13069, https://github.com/microsoft/PowerToys/issues/3959

## Contributor License Agreement (CLA)
A CLA must be signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA.
